### PR TITLE
Add lambda function command

### PR DESCRIPTION
**Pull request by blowekamp**
## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/Core/Common/src/itkObject.cxx [hunk 1]
```diff
@@ -503,6 +503,15 @@ Object ::AddObserver(const EventObject & event, Command * cmd) const
   return this->m_SubjectImplementation->AddObserver(event, cmd);
 }
 
+unsigned long
+Object::AddObserver(const EventObject & event, std::function<void(const EventObject &)> function) const
+{
+  auto cmd = FunctionCommand::New();
```
Comment (dzenanz) [inline Modules/Core/Common/src/itkObject.cxx]: Will `cmd` be deleted once object is deallocated?

**Comment by dzenanz** [comment]
We should also update documentation and examples, as this greatly reduces amount of code needed to track events.

File: Modules/Core/Common/include/itkObject.h [hunk 2]
```diff
@@ -156,6 +156,8 @@ class ITKCommon_EXPORT Object : public LightObject
 
   unsigned long
   AddObserver(const EventObject & event, Command *) const;
+  unsigned long
+  AddObserver(const EventObject & event, std::function<void(const EventObject &)> function) const;
```
Comment (phcerdan) [inline Modules/Core/Common/include/itkObject.h]: Do we need the non-const version?

File: Modules/Core/Common/src/itkObject.cxx (hunk 1 above)
Comment (blowekamp) [reply Modules/Core/Common/src/itkObject.cxx]: Yes the itk::Object holds onto the Commands with SmartPointers as is normal.

**Comment by blowekamp** [comment]
The original callbacks had this signature:
```
void f(const Object * caller, const EventObject & event)
```

Here I removed the object caller parameter to avoid the odd two signatures of the callback and invocation. 

I am considering also removing the event argument too, as that can easily be replaced with lambda captures. Here is usage of the shortened usage:
```
o->AddObserver(itk::AnyEvent(), [&cnt]() { ++cnt; });
```

Any thought if either of those arguments are needed?

File: Modules/Core/Common/include/itkObject.h (hunk 2 above)
Comment (blowekamp) [reply Modules/Core/Common/include/itkObject.h]: That was part of the reason, I removed the object argument to the callback, I was also considering removing the EventObject argument too.

**Comment by dzenanz** [comment]
Listening to iteration events is probably the most common. I would keep the arguments. The compiler will remove them if they are not needed, and they provide generality. If you really want the (short-form) non-argument lambda to be available, you could provide both signatures.

**Comment by blowekamp** [comment]
> Listening to iteration events is probably the most common. I would keep the arguments. The compiler will remove them if they are not needed, and they provide generality. If you really want the (short-form) non-argument lambda to be available, you could provide both signatures.

So for iteration events why would you need the arguments? The EventObjects contain no additional information beyond the type of event, which is known when you add the command. And the object can easily be captured.

With both arguments the inline addition of observer would look like:
```
 o->AddObserver(itk::AnyEvent(), [](Object * caller, const EventObject & event) {std::cout << event.GetEventName() << std::end; });
```

or 

```
auto cmd = [](Object * caller, const EventObject & event)
{
   std::cout << event.GetEventName() << std::end;
 };
o->AddObserver(itk::AnyEvent(), std::move(cmd));
```

I am no doubt biased by what is implemented in SimpleITK:
https://github.com/SimpleITK/SimpleITK/blob/f8d1ff4328dd2f31d132b01e72d86667e98d3a6d/Examples/FilterProgressReporting/FilterProgressReporting.cxx

**Comment by dzenanz** [comment]
I thought you were referring to arguments of `AddObserver` method. The current proposal is a good balance, in case someone wants to handle different event types differently.

**Comment by thewtex** [comment]
@blowekamp thanks for working on this -- looks very cool!

Please refrain from merging until after the 5.1.0 tag too allow for time to iron out build issues and the interface for this important, new feature.

**Comment by dzenanz** [comment]
Can you fix compile issues and rebase? Now is a good time to merge this.

**Pull request merged**